### PR TITLE
restaurant model testing

### PR DESCRIPTION
**Pull request by omer-katzir**
Adding restaurant model RSpec testing.
Subjects: RSpec, FactoryBot, Faker, traits, circleci and env stuff...

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -47,3 +58,20 @@ jobs:
...
+          name: Database setup
+          command: bin/rails db:schema:load --trace
+
+      - run:
+          name: run espec
+          command: |
+            bundle exec rspec
+
+      # Save artifacts
+      - store_test_results:
```
Comment (kobim) [inline .circleci/config.yml:75]: if using this directive, be sure to [follow the instructions on CircleCI](https://circleci.com/docs/2.0/collect-test-data/#enabling-formatters)

File: .rspec [hunk 2]
```diff
@@ -0,0 +1 @@
+--require spec_helper
```
Comment (kobim) [inline .rspec:1]: you might want to require `rails_helper` instead, and avoid the `require 'rails_helper'` in your test file.

File: app/models/restaurant.rb [hunk 3]
```diff
@@ -16,7 +16,8 @@
 #
 
 class Restaurant < ApplicationRecord
-  enumify :cuisine, [:American, :Asian, :Bakery, :Fast_Food, :Steak, :Sushi, :Vegetarian]
+  E_CUISINES = [:American, :Asian, :Bakery, :Fast_Food, :Steak, :Sushi, :Vegetarian].freeze
```
Comment (kobim) [inline app/models/restaurant.rb:18]: why do you need the freeze here? ([hint](https://github.com/omer-katzir/WeEat/blob/9778634eee5f9c2938a35dba41951132c2506683/.rubocop.yml#L16))

File: config/database.yml [hunk 4]
```diff
@@ -17,6 +17,7 @@
 default: &default
   adapter: postgresql
   encoding: unicode
+  username: omer.katzir
```
Comment (kobim) [inline config/database.yml]: this makes the whole project very specific to your environment, I would lose the username here.
If you need it for the tests, make sure to modify the `test` line.

File: spec/models/restaurant_spec.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  context 'invalid operations' do
+    it 'logs error when creating without cuisine' do
+      expect(Restaurant.create(name: 'test').errors.messages[:cuisine]).not_to be_empty
+    end
+    it 'logs error when creating without name' do
+      expect(Restaurant.create(cuisine: 'test').errors.messages[:name]).not_to be_empty
+    end
+  end
+
+  context 'uuid' do
```
Comment (kobim) [inline spec/models/restaurant_spec.rb]: you can omit this test. The DB is responsible for generating the uuid, so you can assume it does a good job doing that.

File: spec/models/restaurant_spec.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    let(:restaurant) { Restaurant.create(name: 'test', cuisine: Restaurant::E_CUISINES[0]) }
+    it 'checks uuid length and chars' do
+      expect(restaurant.id.tr('-', '').length).to eq(32)
+    end
+  end
+end
+
+module Shoulda
```
Comment (kobim) [inline spec/models/restaurant_spec.rb]: don't put 2 modules in the same file.
This can be helper for future models as well, make sure it will be available for all test files

File: spec/models/restaurant_spec.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    it { is_expected.to validate_presence_of(:name) }
+    it { is_expected.to validate_num_in_range(:rating, 0, 3) }
+    it { is_expected.to validate_num_in_range(:max_delivery_time_min, 1, 720) }
+    it { is_expected.to validate_num_in_range(:latitude, -90, 90) }
+    it { is_expected.to validate_num_in_range(:longitude, -180, 180) }
+    it { is_expected.to validate_inclusion_of(:cuisine).in_array(Restaurant::E_CUISINES) }
+  end
+
+  context 'defaults' do
+    let(:restaurant) { Restaurant.create(name: 'test', cuisine: Restaurant::E_CUISINES[0]) }
```
Comment (kobim) [inline spec/models/restaurant_spec.rb]: take a look at [FactoryBot](https://connect.weworkers.io/display/DIGI/Technical+Resources#TechnicalResources-FactoryBot)

File: spec/models/restaurant_spec.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    it 'has default rating of 0' do
+      expect(restaurant.rating).to eq(0)
+    end
+    it 'has default accept10bis value of false' do
+      expect(restaurant.accept10bis).to eq(false)
+    end
+  end
+
+  context 'invalid operations' do
```
Comment (kobim) [inline spec/models/restaurant_spec.rb]: this is redundant - `validate_presense_of` takes care of that

File: spec/models/restaurant_spec.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    it { is_expected.to validate_inclusion_of(:cuisine).in_array(Restaurant::E_CUISINES) }
+  end
+
+  context 'defaults' do
+    let(:restaurant) { Restaurant.create(name: 'test', cuisine: Restaurant::E_CUISINES[0]) }
+    it 'has default rating of 0' do
+      expect(restaurant.rating).to eq(0)
+    end
+    it 'has default accept10bis value of false' do
+      expect(restaurant.accept10bis).to eq(false)
```
Comment (kobim) [inline spec/models/restaurant_spec.rb]: minor: `expect(x).to be_falsey` is the rspec way...

File: spec/models/restaurant_spec.rb (hunk 7 above)
Comment (kobim) [inline spec/models/restaurant_spec.rb]: In tests we prefer as much as randomability as possible. Take a look at the [sample](https://ruby-doc.org/core-2.5.1/Array.html#method-i-sample) method

File: spec/models/restaurant_spec.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
+require 'rails_helper'
```
Comment (kobim) [inline spec/models/restaurant_spec.rb]: [Run annotate](https://github.com/ctran/annotate_models#usage-in-rails) right now, it should generate the model schema here

File: spec/models/restaurant_spec.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+RSpec.describe Restaurant, type: :model do
+  context 'validations' do
+    it { is_expected.to validate_presence_of(:name) }
+    it { is_expected.to validate_num_in_range(:rating, 0, 3) }
+    it { is_expected.to validate_num_in_range(:max_delivery_time_min, 1, 720) }
+    it { is_expected.to validate_num_in_range(:latitude, -90, 90) }
+    it { is_expected.to validate_num_in_range(:longitude, -180, 180) }
+    it { is_expected.to validate_inclusion_of(:cuisine).in_array(Restaurant::E_CUISINES) }
+  end
```
Comment (kobim) [inline spec/models/restaurant_spec.rb:28]: what about `accept10bis`?

**Review by kobim — APPROVED** [review]
🎉!
Check my comments about gems we are using to make testing easier.
Also note you should make your code as agnostic to the environment as possible (hard-coded `omer.katzir` as db username for everyone isn't that good...)

File: app/models/restaurant.rb (hunk 3 above)
Comment (omer-katzir) [reply app/models/restaurant.rb:18]: not sure what you mean, the FrozenStringLiteralComment = disabled, was also in the default rubocop.yml you showed me,
also tried to enable it and place the magic comment in all the relevant files plus removing the freeze call, still got offense on it

File: spec/models/restaurant_spec.rb (hunk 11 above)
Comment (omer-katzir) [reply spec/models/restaurant_spec.rb:28]: got this warning:
Warning from shoulda-matchers:

**You are using `validate_inclusion_of` to assert that a boolean column
allows boolean values and disallows non-boolean ones. Be aware that it
is not possible to fully test this, as boolean columns will
automatically convert non-boolean values to boolean ones. Hence, you
should consider removing this test.**

File: .circleci/config.yml [hunk 12]
```diff
@@ -47,3 +58,30 @@ jobs:
...
+          path: /tmp/test-results
+          steps:
```
Comment (kobim) [inline .circleci/config.yml]: I think this is a mistake - this key is ignored.

File: .rspec (hunk 2 above)
Comment (kobim) [reply .rspec:1]: `rails_helper` includes `spec_helper`, you don't need both

File: .rspec [hunk 13]
```diff
@@ -0,0 +1,5 @@
...
+--require rails_helper
+--require helpers/matcher_helper
```
Comment (kobim) [inline .rspec]: are you sure both lines are required?
Usually, this `.rspec` helper is just to let rspec know which files to load at start. You shouldn't "abuse" it and replace your `require`s with it.

File: Gemfile [hunk 14]
```diff
@@ -49,3 +49,16 @@ gem 'rubocop', require: false
...
+  gem 'shoulda-matchers', '~> 4.0'
+end
+
+group :test do
+  gem 'factory_bot_rails'
+end
+
+gem 'rspec_junit_formatter'
+
+gem 'faker'
```
Comment (kobim) [inline Gemfile]: can faker go to the `test` group as well?

File: Gemfile [hunk 15]
```diff
@@ -49,3 +49,16 @@ gem 'rubocop', require: false
 gem 'active_model_serializers', '~> 0.10.0'
 
 gem 'enumify'
+
+group :development, :test do
+  gem 'rspec-rails', '~> 3.8'
+  gem 'shoulda-matchers', '~> 4.0'
+end
+
+group :test do
```
Comment (kobim) [inline Gemfile:53]: what's the difference between this group and the previous one? seems like both relate to tests.

File: app/models/restaurant.rb [hunk 16]
```diff
@@ -13,10 +15,11 @@
 #  rating                :float            default(0.0)
 #  created_at            :datetime         not null
 #  updated_at            :datetime         not null
-#
+# frozen_string_literal: true
```
Comment (kobim) [inline app/models/restaurant.rb]: not needed twice.

File: app/models/restaurant.rb [hunk 17]
```diff
@@ -1,3 +1,5 @@
...
```
Comment (kobim) [inline app/models/restaurant.rb]: If you decided to use that, you need to use that in all your files.
Remove the rule you ignored in `rubocop` and fix the rest of the files.

File: config/application.rb [hunk 18]
```diff
@@ -24,6 +24,9 @@ class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 6.0
 
+    #  to load factories from custom folder
```
Comment (kobim) [inline config/application.rb]: why are those comments here?

File: config/database.yml [hunk 19]
```diff
@@ -58,6 +58,7 @@ development:
 test:
   <<: *default
   database: we_eat_test
+  username: omer.katzir
```
Comment (kobim) [inline config/database.yml]: should be more generic

File: spec/factories/restaurant_factory.rb [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+#  cuisine               :string
+#  latitude              :float
+#  longitude             :float
+#  max_delivery_time_min :integer
+#  name                  :string           not null
+#  rating                :float            default(0.0)
+#  created_at            :datetime         not null
+#  updated_at            :datetime         not null
+#
+require 'faker'
```
Comment (kobim) [inline spec/factories/restaurant_factory.rb:18]: space before and after `require` please

File: spec/factories/restaurant_factory.rb [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+FactoryBot.define do
+  factory :restaurant do
+    name { Faker::Restaurant.name }
+    cuisine { Restaurant::E_CUISINES.sample }
+  end
+  factory :restaurant_no_cuisine do
```
Comment (kobim) [inline spec/factories/restaurant_factory.rb]: see [Traits](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#traits) for this kind of a behaviour

File: spec/rails_helper.rb [hunk 22]
```diff
@@ -0,0 +1,69 @@
+# This file is copied to spec/ when you run 'rails generate rspec:install'
+require 'spec_helper'
+require 'factory_bot'
```
Comment (kobim) [inline spec/rails_helper.rb]: you don't need this in both `.rspec` and this file. See my comment in `.rspec`

File: spec/support/factory_bot.rb [hunk 23]
```diff
@@ -0,0 +1,5 @@
+require 'factory_bot_rails'
```
Comment (kobim) [inline spec/support/factory_bot.rb]: not sure this file is required
(especially when `rails_helper.rb` already exists with `RSpec.configure do |config|`...)

File: spec/support/factory_bot.rb [hunk 24]
```diff
@@ -0,0 +1,5 @@
...
+RSpec.configure do |config|
+  config.include FactoryBot::Syntax::Methods
+  FactoryBot.definition_file_paths = [File.expand_path('factories', __dir__)]
```
Comment (kobim) [inline spec/support/factory_bot.rb]: you need this only
> If you're using a separate directory structure for your factories

([GETTING_STARTED](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#using-without-bundler))

File: app/models/restaurant.rb (hunk 17 above)
Comment (omer-katzir) [reply app/models/restaurant.rb]: this is a leftover from the test I've done enabling the magic comment cop... 
removing the comment

File: spec/factories/restaurant_factory.rb [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+require 'faker'
+
+FactoryBot.define do
+  factory :restaurant, traits: [:name, :cuisine] do
+    trait :name do
+      name { Faker::Restaurant.name }
+    end
+    trait :cuisine do
+      cuisine { Restaurant::E_CUISINES.sample }
```
Comment (kobim) [inline spec/factories/restaurant_factory.rb]: traits are used to show different states of the object. You should usually apply all the properties as default (name, cuisine) and have traits override the ones you don't want.

File: spec/factories/restaurant_factory.rb [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+#  name                  :string           not null
+#  rating                :float            default(0.0)
+#  created_at            :datetime         not null
+#  updated_at            :datetime         not null
+#
+
+require 'faker'
+
+FactoryBot.define do
+  factory :restaurant, traits: [:name, :cuisine] do
```
Comment (kobim) [inline spec/factories/restaurant_factory.rb]: not sure you need to define the traits here.

Comment (kobim) [inline spec/factories/restaurant_factory.rb]: what about the other values? you should generate them all, as it will help you in the future.

File: spec/controllers/restaurants_controller_spec.rb [hunk 27]
```diff
@@ -0,0 +1,10 @@
+RSpec.describe RestaurantsController, type: :controller do
```
Comment (kobim) [inline spec/controllers/restaurants_controller_spec.rb]: this PR is model testing, not controller testing. Please remove this.

File: .circleci/config.yml [hunk 28]
```diff
@@ -47,3 +58,21 @@ jobs:
...
+      - run:
+          name: Wait for DB
+          command: dockerize -wait tcp://localhost:5432 -timeout 1m
+
+      - run:
+          name: Database setup
+          command: bin/rails db:schema:load --trace
+
+      - run:
+          name: run espec
```
Comment (kobim) [inline .circleci/config.yml]: typo: `espec` instead of `rspec`

File: .circleci/config.yml [hunk 29]
```diff
@@ -10,16 +10,27 @@ jobs:
...
+          PGUSER: omer.katzir
+          RAILS_ENV: test
 
       # Specify service dependencies here if necessary
       # CircleCI maintains a library of pre-built images
       # documented at https://circleci.com/docs/2.0/circleci-images/
-      # - image: circleci/postgres:9.4
+      - image: circleci/postgres:9.4
+        environment:
+          POSTGRES_USER: omer.katzir
```
Comment (kobim) [inline .circleci/config.yml]: should be more generic.

File: .circleci/config.yml [hunk 30]
```diff
@@ -10,16 +10,27 @@ jobs:
...
+          POSTGRES_DB: we_eat_test
+          POSTGRES_PASSWORD: ""
```
Comment (kobim) [inline .circleci/config.yml:24]: can this be omitted?

File: .circleci/config.yml [hunk 31]
```diff
@@ -10,16 +10,27 @@ jobs:
       - image: circleci/ruby:2.5.1-node
         environment:
           BUNDLER_VERSION: 2.0.1
+          PGHOST: 127.0.0.1
+          PGUSER: omer.katzir
+          RAILS_ENV: test
```
Comment (kobim) [inline .circleci/config.yml:15]: out of curiosity - isn't this the default when running inside CircleCI image?

File: spec/support/factory_bot.rb [hunk 32]
```diff
@@ -0,0 +1,4 @@
+RSpec.configure do |config|
```
Comment (kobim) [inline spec/support/factory_bot.rb]: is this file required?

Comment (omer-katzir) [reply spec/support/factory_bot.rb]: [https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#rspec](url)

File: .circleci/config.yml (hunk 31 above)
Comment (omer-katzir) [reply .circleci/config.yml:15]: tried running without these fields and failed on DB creation

File: .circleci/config.yml (hunk 30 above)
Comment (omer-katzir) [reply .circleci/config.yml:24]: user name can be blank but fields need to exist otherwise it fails

File: .circleci/config.yml (hunk 29 above)
Comment (omer-katzir) [reply .circleci/config.yml]: clearing it does the trick?

File: spec/support/factory_bot.rb (hunk 32 above)
Comment (kobim) [reply spec/support/factory_bot.rb]: what if you add this in `rails_helper.rb`?

File: spec/rails_helper.rb [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+  # https://relishapp.com/rspec/rspec-rails/docs
+  config.infer_spec_type_from_file_location!
+
+  # Filter lines from Rails gems in backtraces.
+  config.filter_rails_from_backtrace!
+  # arbitrary gems may also be filtered via:
+  # config.filter_gems_from_backtrace("gem name")
+
+  config.include FactoryBot::Syntax::Methods
+  FactoryBot.definition_file_paths = [File.expand_path('factories', __dir__)]
```
Comment (kobim) [inline spec/rails_helper.rb]: I think this line can be removed

File: spec/factories/restaurant_factory.rb [hunk 34]
```diff
@@ -0,0 +1,51 @@
...
+#
+
+require 'faker'
+
+FactoryBot.define do
+  factory :restaurant do
+    name { Faker::Restaurant.name }
+    cuisine { Restaurant::E_CUISINES.sample }
+    trait :accept10bis do
+      accept10bis { [true, false].sample }
```
Comment (kobim) [inline spec/factories/restaurant_factory.rb]: this shouldn't be in a trait

File: spec/factories/restaurant_factory.rb [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+    end
+
+    trait :with_location do
```
Comment (kobim) [inline spec/factories/restaurant_factory.rb:28]: 🍾

File: spec/factories/restaurant_factory.rb [hunk 36]
```diff
@@ -0,0 +1,51 @@
...
+      latitude { addr.latitude }
+      longitude { addr.longitude }
+    end
+
+    trait :max_delivery_time_min do
+      max_delivery_time_min { [1..720].sample }
+    end
+
+    trait :rating do
+      rating { [0.0..3.0].sample }
```
Comment (kobim) [inline spec/factories/restaurant_factory.rb]: this shouldn't be in a trait

File: spec/factories/restaurant_factory.rb [hunk 37]
```diff
@@ -0,0 +1,51 @@
...
+
+    trait :with_location do
+      addr = Faker::Address
+      address { addr.full_address }
+      latitude { addr.latitude }
+      longitude { addr.longitude }
+    end
+
+    trait :max_delivery_time_min do
+      max_delivery_time_min { [1..720].sample }
```
Comment (kobim) [inline spec/factories/restaurant_factory.rb]: this shouldn't be in a trait

File: spec/factories/restaurant_factory.rb (hunk 35 above)
Comment (omer-katzir) [reply spec/factories/restaurant_factory.rb:28]: ?

File: .rspec (hunk 2 above)
Comment (kobim) [reply .rspec:1]: bump ^^

File: spec/factories/restaurant_factory.rb (hunk 36 above)
Comment (kobim) [reply spec/factories/restaurant_factory.rb]: (but should be present.)

File: spec/factories/restaurant_factory.rb (hunk 37 above)
Comment (kobim) [reply spec/factories/restaurant_factory.rb]: (but should be present.)
